core/new-gui/src/app/workspace/component/code-editor-dialog/code-editor-dialog.component.ts

```diff
@@ -56,9 +61,47 @@ export class CodeEditorDialogComponent implements AfterViewInit, SafeStyle, OnDe
 
   ngOnDestroy(): void {
     this.workflowActionService.getTexeraGraph().updateSharedModelAwareness("editingCode", false);
+    if (this.socket) {
+      this.socket.close();
+      this.socket = undefined;
+    }
+  }
+
+  createLanguageClient(transports: MessageTransports): MonacoLanguageClient {
+    return new MonacoLanguageClient({
+      name: "Python UDF Language Client",
+      clientOptions: {
+        documentSelector: ["python"],
+        errorHandler: {
+          error: () => ({ action: ErrorAction.Continue }),
+          closed: () => ({ action: CloseAction.DoNotRestart }),
```

Comment:
what is the closed() callback? Can it be used so that  when the websocket gets closed unexpectly it may be able to reconnect?

Reply:
Hi, it is a function that gets executed when the connection between the language client and server is closed.

This behavior is useful in cases where the language server is terminated unexpectedly or intentionally, and the client does not need to attempt to reconnect to the server by returning `CloseAction.DoNotRestart`.

Reply:
when terminated intentionally, we do not need to restart. however, when terminated unexpectedly, wouldn't it be better to restart if possible?

Reply:
I think using CloseAction.DoNotRestart in the closed() callback has the advantage of avoiding an endless cycle of reconnection attempts in cases where the language server is unavailable or cannot be reached. Overall I think Restart makes sense In situations where the language server is known to be temporarily unavailable.